spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/test/AbstractProcessDeployerTests.java

```diff
@@ -0,0 +1,253 @@
+				Matchers.<ProcessStatus>hasProperty("state", is(unknown))), timeout.noAttempts, timeout.pause));
+
+	}
+
+	/**
+	 * Tests that a module which takes a long time to deploy is correctly reported as deploying.
+	 * Test that such a module can be killed (undeployed).
+	 */
+	@Test
+	public void testDeployingStateCalculationAndCancel() {
+		Map<String, String> properties = new HashMap<>();
+		properties.put("initDelay", "" + 1000 * 60 * 60); // 1hr
+		AppDefinition definition = new AppDefinition(randomName(), properties);
+		Resource resource = mavenResource("org.springframework.cloud.stream.module", "integration-test-processor","jar","exec","1.0.0.BUILD-SNAPSHOT");
+		AppDeploymentRequest request = new AppDeploymentRequest(definition, resource, properties);
+
+		String deploymentId = processDeployer().deploy(request);
+		Attempts timeout = deploymentTimeout();
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(deploying))), timeout.noAttempts, timeout.pause));
+
+		timeout = undeploymentTimeout();
+		processDeployer().undeploy(deploymentId);
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(unknown))), timeout.noAttempts, timeout.pause));
+
+	}
+
+	@Test
+	public void testFailedDeployment() {
+		Map<String, String> properties = new HashMap<>();
+		properties.put("killDelay", "0");
+		AppDefinition definition = new AppDefinition(randomName(), properties);
+		Resource resource = mavenResource("org.springframework.cloud.stream.module", "integration-test-processor","jar","exec","1.0.0.BUILD-SNAPSHOT");
+		AppDeploymentRequest request = new AppDeploymentRequest(definition, resource, properties);
+
+		String deploymentId = processDeployer().deploy(request);
+		Attempts timeout = deploymentTimeout();
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(failed))), timeout.noAttempts, timeout.pause));
+
+		timeout = undeploymentTimeout();
+		processDeployer().undeploy(deploymentId);
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(unknown))), timeout.noAttempts, timeout.pause));
+	}
+
+	protected String randomName() {
+		return UUID.randomUUID().toString();
+	}
+
+	/**
+	 * Return the timeout to use for repeatedly querying a module while it is being deployed.
+	 * Default value is one minute, being queried every 5 seconds.
+	 */
+	protected Attempts deploymentTimeout() {
+		return new Attempts(12, 5000);
+	}
+
+	/**
+	 * Return the timeout to use for repeatedly querying a module while it is being un-deployed.
+	 * Default value is one minute, being queried every 5 seconds.
+	 */
+	protected Attempts undeploymentTimeout() {
+		return new Attempts(20, 5000);
+	}
+
+	private Resource mavenResource(String groupId, String artifactId, String ext, String classifier, String version) {
+		// TODO: actually resolve using Aether,
```

Comment:
this could use the `MavenResource` that is part of this project (it needs some work, but currently it resolves from the local repo only)